src/Heap.cpp

```diff
@@ -0,0 +1,295 @@
+			fbHead = cur->getNextFree();
+		} else {
+			// cur is not the head of the free list, set the next free block as the next free block of the previous block
+			*(FreeBlock**) prev->getNextFreePointer() = cur->getNextFree();
+		}
+	} else {
+#if DEBUG
+		std::cout << "Splitting free block!" << std::endl;
+#endif
+		// Split
+		// Create a new FreeBlock after the allocated block
+		FreeBlock* newFreeBlock = new ((char*) cur + requestedTotalSize) FreeBlock(leftOver - (int) sizeof(FreeBlock));
+		// Set the next freeBlock of the new FreeBlock to the next free block of cur
+		newFreeBlock->setNextFreePointer(cur->getNextFree());
+		// Set the nextFreeBlock of prev to the new FreeBlock
+		if (prev == nullptr) {
+			// cur is the head of the free list, set the new FreeBlock as the new head
+			fbHead = newFreeBlock;
+		} else {
+			// cur is not the head of the free list, set the new FreeBlock as the next free block of the previous block
+			*(FreeBlock**) prev->getNextFreePointer() = newFreeBlock;
+		}
+	}
+
+	// Set the typeDescriptor of the allocated block
+	Block* block = new (cur) Block(typeDescriptor);
+
+	// Adjust free_bytes
+	free_bytes -= requestedTotalSize;
+
+	// Return the allocated block
+	return block;
+}
+
+/*
+// Commented out, as I think this is not actually needed.
+// The GC should perform deallocate of unused blocks in a more general context.
+void Heap::dealloc(Block* block) {
+	std::cout << "Deallocating memory..." << std::endl;
+
+	auto p = (Block*) heap_buffer;
+	Block* left = nullptr;
+	while (p != block) {
+		left = p;
+		p = (Block*) ((char*) p + p->dataSize() + sizeof(Block));
+	}
+	if (left != nullptr && !left->used) {
+		// merge left
+	} else {
+		// add p to freelist
+	}
+
+	Block* right = (Block*) ((char*) p + p->dataSize() + (int) sizeof(Block));
+	if (!right->used) {
+		// remove right from freelist
+		// merge p and right
+	}
+
+	// Implement the reallocation logic here
+}
+*/
+
+/// Register a typeDescriptor with the Heap. Will return true if the typeDescriptor was successfully registered, false otherwise.
+bool Heap::registerType(const std::string& type, TypeDescriptor& descriptor) {
+	if (type_map.find(type) != type_map.end()) {
+		std::cout << "Type " << type << " already exists!" << std::endl;
+		return false;
+	}
+	std::cout << "Registering typeDescriptor " << type << std::endl;
+	type_map[type] = &descriptor;
+	return true;
+}
+
+std::string Heap::ToString() {
+	std::string postfix;
+	std::string str = "Heap {\n";
+	str += "Heap dataSize: " + std::to_string(HEAP_SIZE) + "\n";
+	str += "Free bytes: " + std::to_string(free_bytes) + " | Used bytes: " + std::to_string((HEAP_SIZE - free_bytes)) + "\n";
+	str += "Stored typeDescriptor descriptors: " + std::to_string(type_map.size()) + "\n";
+
+	str += "Live blocks { ";
+	int traversedSize = 0;
+	Block* bCur = (Block*) heap_buffer;
+	while (bCur != nullptr && traversedSize < HEAP_SIZE) {
+		int curBlkSize;
+		if (bCur->isFreeBlock()) {
```

Comment:
isn't this the other way round?
```suggestion
		if (!bCur->isFreeBlock()) {
```

Reply:
Jep, I just replaced `used` with `isFreeBlock()` and forgot the negation